docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,841 @@
+
+Environment providers like Workerd, can expose an environment configurator for the most common case of using the same runtime for both dev and build environments. The default environment options can also be set so the user doesn't need to do it.
+
+```js
+function createWorkedEnvironment(userConfig) {
+  return mergeConfig(
+    {
+      resolve: {
+        conditions: [
+          /*...*/
+        ],
+      },
+      dev: {
+        createEnvironment: (server, name) =>
+          createWorkerdDevEnvironment(server, name),
+      },
+      build: {
+        createEnvironment: (builder, name) =>
+          createWorkerdBuildEnvironment(builder, name),
+      },
+    },
+    userConfig,
+  )
+}
+```
+
+Then the config file can be writen as
+
+```js
+import { workerdEnvironment } from 'vite-environment-workerd'
+
+export default {
+  environments: {
+    ssr: createWorkerdEnvironment({
+      build: {
+        outDir: '/dist/ssr',
+      },
+    }),
+    rsc: createWorkerdEnvironment({
+      build: {
+        outDir: '/dist/rsc',
+      },
+    }),
+  ],
+}
+```
+
+In this case we see how a Workerd environment can be set for both the default SSR environment and for a new custom RSC environment.
+
+## Plugins and environments
+
+### Accessing the current environment in hooks
+
+The Vite server has a shared plugin pipeline, but when a module is processed it is always done in the context of a given environment. The `environment` instance is available in the plugin context of `resolveId`, `load`, and `transform`.
+
+A plugin could use the `environment` instance to:
+
+- Only apply logic for certain environments.
+- Change the way they work depending on the configuration for the environment, which can be accessed using `environment.config`. The vite core resolve plugin modifies the way it resolves ids based on `environment.config.resolve.conditions` for example.
+
+```ts
+  transform(code, id) {
+    console.log(this.enviroment.config.resolve.conditions)
+  }
+```
+
+### Registering new environments using hooks
+
+Plugins can add new environments in the `config` hook:
+
+```ts
+  config(config: UserConfig) {
+    config.environments.rsc ??= {}
+  }
+```
+
+An empty object is enough to register the environment, default values from the root level environment config.
+
+### Configuring environment using hooks
+
+While the `config` hook is running, the complete list of environments isn't yet known and the environments can be affected by both the default values from the root level environment config or explicitely through the `config.environments` record.
+Plugins should set default values using the `config` hook. To configure each environment, they can use the new `configEnvironment` hook. This hook is called for each environment with its partially resolved config including resolution of final defaults.
+
+```ts
+  configEnvironment(name: string, config: EnvironmentConfig) {
+    if (name === 'rsc') {
+      config.resolve.conditions = // ...
+```
```

Comment:
I don't think we should promote modifying the config in `configResolved`. When this hook is called, it should be safe to assume that the config is settled and plugins can read it and cache/store what they need. If we want to do this config in another hook, it should be in the `config` hook. And we could do that without a problem. The only reason why the `configEnvironment` hook exists is to help plugins deal with the complexity of environment overrides. I'm ok removing it if we think it is redundant, but at this point I feel that it will be very helpful to avoid subtle bugs in plugins.

Reply:
Is this `configEnvironment` only for custom environments or do you also want to move people to configure `name = "client"` and `name = "ssr"` cases here as well? For example,

```ts
// from this
config(config, env) {
  return {
    optimizeDeps: { ... },
    ssr: { optimizeDeps: { ...} }
  }
}

// to this?
configEnvironment(name, config) {
  if (name === "client") {
    config.dev.optimizeDeps = { ... }
  }
  if (name === "ssr") {
    config.dev.optimizeDeps = { ... }
  }
}

// or this?
config(config, env) {
  config.environments.client.dev.optimizeDeps = { ... }
  config.environments.ssr.dev.optimizeDeps = { ... }
}

// or this?
config(config, env) {
  return {
    environments: {
      client: {
        dev: {
          optimizeDeps: { ... }
        }
      }
      ssr: { ... }
    }
  }
}
```

I thought it will be mostly redundant, but after writing this sample code, maybe `configEnvironment` looks explicit and nicer even for builtin ones. But maybe this assignment form of config needs to take care config mergining inside the hook itself?

Reply:
Ah, I forgot that `ResolvedConfig` is a readonly. 😅

> At this point I feel that it will be very helpful to avoid subtle bugs in plugins.

What kind of bug do you have in mind?

Reply:
> Is this `configEnvironment` only for custom environments or do you also want to move people to configure `name = "client"` and `name = "ssr"` cases here as well?

Yes, `configEnvironment` will be called once per environment (including default ones).

Reply:
> But maybe this assignment form of config needs to take care config merging inside the hook itself?

In my comment above, I was worrying about how merging works in `configEnvironment`, but I just realized it can return partial options and merged automatically.

```ts
    configEnvironment(name, _config, _env) {
      if (name === "ssr") {
        return {
          build: {
            rollupOptions: {
              input: {
                index: entry,
              },
            },
          },
        };
      }
    },
```